url088e9d87be_jpush_jpush_java_library_tgz-pJ8-cn_jpush_example_JPushClientExampleJ8/src/cn/jpush/api/JPushClient.java

```diff
@@ -0,0 +1,299 @@
+package cn.jpush.api;
+
+import java.util.Map;
+
+/**
+ * The entrance of JPush API library.
+ *
+ */
+public class JPushClient extends BaseClient {
+
+	protected static HttpPostClient httpClient = new HttpPostClient();
+
+	public JPushClient(String masterSecret, String appKey) {
+		this.masterSecret = masterSecret;
+		this.appKey = appKey;
+	}
+
+	public JPushClient(String masterSecret, String appKey, long timeToLive) {
+		this.masterSecret = masterSecret;
+		this.appKey = appKey;
+		this.timeToLive = timeToLive;
+	}
+
+	public JPushClient(String masterSecret, String appKey, DeviceEnum device) {
+		this.masterSecret = masterSecret;
+		this.appKey = appKey;
+		devices.add(device);
+	}
+
+	public JPushClient(String masterSecret, String appKey, long timeToLive, DeviceEnum device) {
+		this.masterSecret = masterSecret;
+		this.appKey = appKey;
+		this.timeToLive = timeToLive;
+		this.devices.add(device);
+	}
+	
+	/*
+	 * @description 发送带TAG的通知
+	 * @return MessageResult
+	 */
+	public MessageResult sendNotificationWithTag(int sendNo, String tag, String msgTitle, String msgContent) {
+		NotifyMessageParams p = new NotifyMessageParams();
+		p.setReceiverType(ReceiverTypeEnum.TAG);
+		p.setReceiverValue(tag);
+		return sendNotification(p, sendNo, msgTitle, msgContent, 0, null);
+	}
+
+	/*
+	 * @params builderId通知栏样式
+	 * @description 发送带TAG的通知
+	 * @return MessageResult
+	 */
+	public MessageResult sendNotificationWithTag(int sendNo, String tag, String msgTitle, String msgContent, int builderId, Map<String, Object> extra) {
+		NotifyMessageParams p = new NotifyMessageParams();
+		p.setReceiverType(ReceiverTypeEnum.TAG);
+		p.setReceiverValue(tag);
+		return sendNotification(p, sendNo, msgTitle, msgContent, builderId, extra);
+	}
+
+	/*
+	 * @params overrideMsgId 待覆盖的上一条消息的 ID
+	 * @params builderId通知栏样式
+	 * @description 发送带TAG的通知
+	 * @return MessageResult
+	 */
+	public MessageResult sendNotificationWithTag(int sendNo, String tag, String msgTitle, String msgContent, int builderId, Map<String, Object> extra,String overrideMsgId) {
+		NotifyMessageParams p = new NotifyMessageParams();
+		p.setReceiverType(ReceiverTypeEnum.TAG);
+		p.setReceiverValue(tag);
+		p.setOverrideMsgId(overrideMsgId);
+
+		return sendNotification(p, sendNo, msgTitle, msgContent, builderId, extra);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 67:72, 105:110, 143:148, 181:186.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.